politeiawww/database/cockroachdb/cockroachdb.go

```diff
@@ -0,0 +1,528 @@
+	if err != nil {
+		log.Debugf("Open: could not connect to %v", c.dbAddress)
+		return err
+	}
+
+	c.usersdb = db
+
+	// See if we need to write a version record.
+	_, err = c.Get(database.DatabaseVersionKey)
+
+	if err == database.ErrNotFound {
+		// Write version record.
+		payload, err := database.EncodeVersion(database.Version{
+			Version: database.DatabaseVersion,
+			Time:    time.Now().Unix(),
+		})
+		if err != nil {
+			return err
+		}
+
+		return c.Put(database.DatabaseVersionKey, payload)
+	} else if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Close shuts down the database.  All interface functions MUST return with
+// errShutdown if the backend is shutting down.
+func (c *cockroachdb) Close() error {
+	log.Tracef("Close")
+
+	c.Lock()
+	defer c.Unlock()
+
+	c.shutdown = true
+	return c.usersdb.Close()
+}
+
+// CreateCDB uses the CockroachDB root user to create a database,
+// politeiawww user if it does not already exist. User permissions are then
+// set for the database and the database tables are created if they do not
+// already exist.
+func CreateCDB(host, net, rootCert, certDir string) error {
+	log.Tracef("Create cockroachDB: %v %v %v %v", host, net, rootCert, certDir)
+	// Connect to CockroachDB as root user. CockroachDB connects
+	// to defaultdb when a database is not specified.
+	h := "postgresql://root@" + host
+	u, err := url.Parse(h)
+	if err != nil {
+		log.Debugf("Create: could not parse url %v", h)
+		return err
+	}
+
+	qs := buildDBQueryString(rootCert, certDir, u)
+
+	addr := u.String() + "?" + qs
+
+	db, err := gorm.Open("postgres", addr)
+	if err != nil {
+		log.Debugf("Create: could not connect to %v", addr)
+		return err
+	}
+	defer db.Close()
+
+	// Setup politeiawww database and users.
+	dbName := dbPrefix + net
+	q := "CREATE DATABASE IF NOT EXISTS " + dbName
+	err = db.Exec(q).Error
+	if err != nil {
+		return err
+	}
+
+	q = "CREATE USER IF NOT EXISTS " + UserPoliteiawww
+	err = db.Exec(q).Error
+	if err != nil {
+		return err
+	}
+	q = "GRANT ALL ON DATABASE " + dbName + " TO " + UserPoliteiawww
+	err = db.Exec(q).Error
+	if err != nil {
+		return err
+	}
+
+	// Connect to records database with root user.
+	h = "postgresql://root@" + host + "/" + dbName
+	u, err = url.Parse(h)
+	if err != nil {
+		log.Debugf("Create: could not parse url %v", h)
+		return err
+	}
+	addr = u.String() + "?" + qs
+	pdb, err := gorm.Open("postgres", addr)
+	defer pdb.Close()
+	if err != nil {
+		log.Debugf("Create: could not connect to %v", addr)
+		return err
+	}
+
+	// Setup database tables.
+	tx := pdb.Begin()
+	err = createTables(tx)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	return tx.Commit().Error
+}
+
+// NewCDB returns a new cockroachdb context which contains a connection to the
+// specified database that was made using the passed in user and certificates.
+func NewCDB(user, host, net, rootCert, certDir string, dbKey *database.EncryptionKey, cfg *Config) (*cockroachdb, error) {
```

Comment:
Style nit.  This function gets called as `cockroachdb.NewCDB()`.  The `CDB` is redundant since the package name is already cockroachdb.

https://golang.org/doc/effective_go.html#package-names